### ISSUE TITLE
DataPreprocessing / Feature Visualization / Trained Model

### ISSUE THREAD

**Issue by samieti02** [Kamnitsask/deepmedic#123]
Hello dear Sir,

First of all, i have a question regarding the data preprocessing. When you say 

> The images of each subject should have the same dimensions (per subject, no need for whole database). This is, the number of voxels per dimension must be the same for all images of a subject

Do it mean, for example, we have three images, A, B and C that ?
`
Xa=Xb=Xc = 200 , Ya=Yb=Yc = 180 , Za=Zb=Zc = 150 `
or 
`
Xa=Xb=Xc=Ya=Yb=Yc=Za=Zb=Zc  = 200
`
?

2) I am interested , on how the network visualizes the features. How can we integrate it in the tool, or which py files are better suitable for that. Can you make us some suggestions? It would be very great.

3) Is it possible to work with networks (from Deepmedic) which are already trained?

Thanks
Sami

**Comment by Kamnitsask** [comment]
Hello Sami, sorry for the very delayed response, too busy schedule lately and I have a hard time following everything, doing my best.

1. On your first question, the answer is the first one.
The dimensions Xa=Xb=Xc, Ya=Yb=Yc. No need for the 3 axis to have the same dims.

2. There is a built-in functionality to visualise the feature maps, to get figures similar to the one in the discussion section of our MedIA paper.

To do that, look in the end of the file: ./examples/config/deepmedic/test/testConfig.cfg
Uncomment the lines for the variables "minMaxIndicesOfFmsToSaveFromEachLayerOfNormalPathway" etc.
Then, specify indices of which feature maps, for which layer, of which pathway you want visualized.
E.g:
minMaxIndicesOfFmsToSaveFromEachLayerOfNormalPathway = [[],[0,999],[],[],[],[],[],[]]
will visualize all (from 0 up to 999) FMs of the second layer of the normal-resolution pathway.

Results will be saved in examples/output/predictions/name-of-train-session/features/

3. I refrain from posting pretrained models, because to make them work, it means that the user would have to correctly apply "exactly" the same preprocessing to his/her test data as I had done when training the pretrained model. Otherwise performance will be bad. I think this is practically not very easy and will get various people confused/pondering. Instead, I believe letting the user train it themselves on their data is not too difficult, and in this way the familiarize themselves with the data and the model too.

That being said, quite some users ask for it, so I am thinking in the back of my head about scenarios where this would be "really" useful but it's not in my high-priority plans.

Thank you,
Kostas

**Comment by samieti02** [comment]
Hello Kostas,
thanks for the reply.

Regarding the visualization, can i have a bit details, about how it works and where to find it in code? The method which is used? I want to cite that too.

For the trained model, it would be greet. Because we dont have as much as data as in brats. I also wanted to test the data on a trained model, to see how good or bad the results are.

Thanks
Sami

**Comment by samieti02** [comment]
Hello Sir, 

please can we have a bit more details about how the visualization works? Which method do you use?

Thanks 
Sami

**Comment by smaritem** [comment]
Hello Sir,

I did not want to start a new topic, because one has already been started. I join the others in this one. I would like to know how it works with the feature visualization. What is your approach? Can you explain it please? Or is it found in publications or articles? Since I'm not a big expert in Python, I can not understand the code that well. But a little explanations would be good.

![fm](https://user-images.githubusercontent.com/22933667/46003784-979a0d00-c0b1-11e8-90f0-3475c4646d1a.PNG)

In the image above, I have saved the FM1 from levels 1 to 4. But I still find it difficult to interpret the images because I don't know exactly how the visualization method works. In most cases, you can see in the plane, feature, like edges. But here I find it difficult to interpret it. Maybe you can give us some hints.

Regarding a pre-trained model, I would also be interested to know what results when testing the data with new trained and pre-trained model and comparing the results

Thanks
Armi

**Comment by Kamnitsask** [comment]
Hello and sorry for the late reply.

There is no special method for visualising the feature maps. All you see is the activations that you get in that layer. The only "technique" is that because because you cannot always apply the network with the whole scan as input (may be possible for a small brain scan, but not for larger images), you need to tile the image into segments/subvolumes, and then stitch the results together. Nothing more.

For all purposes, imagine that the network
 
The "edge detectors" that you may expect is a very simple feature. They may or may not arise in some place in the 1st layer by themselves, somewhere among the many feature maps. Perhaps they may not arise by themselves (they are too simple), and you may see more complicated features even in Layer1. No guarantees. Features in general are not interpretable in the general case, except if you enforce them to be. But that's another discussion. With visual inspection you may or may not find interpretable features. No guarantees.

Regarding pre-trained model, this is not going to happen yet for reasons previously discussed. Sorry for this.

Hope this helps.
Kostas
